### PR TITLE
Added convenience functions for 1 nearest neighbor

### PR DESCRIPTION
**Pull request by BioTurboNick**
If a user wants a single nearest neighbor, unwrapping the return values is a bit cumbersome, when this can be transparently handled by the package instead.

I've added `nn` functions for this purpose, which pass the data along to `knn` with `k = 1` function and unwraps the results.

Also added tests for same.

Open to renaming them `knn` for consistency. EDIT: But I guess that would make the result type-unstable, and thus shouldn't be done.

### CODE REVIEW

**Comment by BioTurboNick** [comment]
I'd like to request integration of this PR.

**Comment by BioTurboNick** [comment]
The `sortres` argument has no effect for k = 1, correct?

File: src/knn.jl [hunk 1]
```diff
@@ -56,17 +56,17 @@ function knn(tree::NNTree{V}, point::Matrix{T}, k::Int, sortres=false, skip::Fun
     knn(tree, new_data, k, sortres, skip)
 end
 
-function nn(tree::NNTree{V}, points::Vector{T}, sortres=false, skip::Function=always_false) where {V, T <: AbstractVector}
-    idx, dist = knn(tree, points, 1, sortres, skip)
+function nn(tree::NNTree{V}, points::Vector{T}, skip::Function=always_false) where {V, T <: AbstractVector}
+    idx, dist = knn(tree, points, 1, skip)
```
Comment (bhgomes) [inline src/knn.jl]: You still have to put `false` here for `sortres` as `knn` needs it.

**Comment by bhgomes** [comment]
There should be an option for looking for the neighbors of a point that you know is already in the dataset used to construct the tree. In that case, we would want `k=2` and select the second element from each list not the first, because the first is the point itself.

```julia
data = [0.861652  0.714498  0.860377  0.178481  0.377927  0.794543  0.872945  0.536459
        0.575687  0.174855  0.25288   0.581422  0.789739  0.632018  0.987797  0.409914]

knn(KDTree(data), data[:,3], 1, true)
# ([3], [0.0])

knn(KDTree(data), data[:,3], 2, true)
# ([3, 2], [0.0, 0.165435])
```

**Comment by BioTurboNick** [comment]
@bhgomes - I don't know how useful that would be, but that isn't relevant to this PR.

File: src/knn.jl (hunk 1 above)
Comment (BioTurboNick) [reply src/knn.jl]: I don't think so. `knn` provides the default of false.

**Comment by bhgomes** [comment]
Okay, point taken.

File: src/knn.jl (hunk 1 above)
Comment (bhgomes) [reply src/knn.jl]: But the order of arguments is important. The test suite fails because the `skip` argument is being read as the `sortres` argument and then there is a boolean interpretation error.

Comment (BioTurboNick) [reply src/knn.jl]: Ah, confused keyword args and successive optional args. Thanks.

**Comment by BioTurboNick** [comment]
@KristofferC - The only failing tests were one in `knn_monkey` in one build of AppVeyor. I don't think this is due to anything I touched, but I'm not sure. Edge case found?

**Comment by m-wells** [comment]
This doesn't seem like the most performant solution to the nearest neighbor case.  When doing `knn` you return a `Vector{Vector{Int}}` but to get `nn` you only need `Vector{Int}`.  Yes I see that this returns `Vector{Int}` but all the work is being done with the `knn` overhead of `Vector{Vector{Int}}`.

**Comment by BioTurboNick** [comment]
@m-wells - Perhaps, but that's not the goal of this PR. Even so, I'm not certain how much overhead that adds and whether attempting to remove the overhead would be worth the effort and added complexity of maintaining a separate algorithm.

**Comment by m-wells** [comment]
@BioTurboNick - fair enough.  I just wanted to point that out.  I think this PR is useful for sure and if an optimization of `nn` ever took place it could be a slide in for these convenience functions.  I took a look to see if it was something that I could slap together but it looks like each tree would need a `nn_kernel` and I don't have that sort of time right now.

**Comment by BioTurboNick** [comment]
Cleaned things up a bit, checks now pass. @KristofferC does it look all set?

**Comment by BioTurboNick** [comment]
Changed to Abstract____ method and tests to match the master branch.

**Comment by BioTurboNick** [comment]
@KristofferC , if this looks okay, may I request this be merged?

**Comment by BioTurboNick** [comment]
Just updated to ensure compatibility with current master. @KristofferC can this be integrated if it is acceptable?
